engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala

```diff
@@ -1083,6 +1048,84 @@ class IrToTruffle(
                 )
               )
           }
+        case Pattern.Type(varName, tpeName, location, _, _) =>
+          tpeName.getMetadata(Patterns) match {
+            case None =>
+              Left(BadPatternMatch.NonVisibleType(tpeName.name))
+            case Some(
+                  BindingsMap.Resolution(BindingsMap.ResolvedType(mod, tpe))
+                ) =>
+              // Using .getTypes because .getType may return an associated type
+              Option(
+                mod.unsafeAsModule().getScope.getTypes.get(tpe.name)
+              ) match {
+                case Some(tpe) =>
+                  val argOfType = List(
+                    IR.DefinitionArgument.Specified(
+                      varName,
+                      None,
+                      None,
+                      suspended = false,
+                      location,
+                      passData    = varName.passData,
+                      diagnostics = varName.diagnostics
+                    )
+                  )
+
+                  val branchCodeNode = childProcessor.processFunctionBody(
+                    argOfType,
+                    branch.expression,
+                    branch.location
+                  )
+                  Right(
+                    CatchTypeBranchNode.build(tpe, branchCodeNode.getCallTarget)
+                  )
+                case None => Left(BadPatternMatch.NonVisibleType(tpeName.name))
+              }
+            case Some(
+                  BindingsMap.Resolution(
+                    BindingsMap.ResolvedPolyglotSymbol(mod, symbol)
+                  )
+                ) =>
+              Option(
+                mod
+                  .unsafeAsModule()
+                  .getScope
+                  .getPolyglotSymbols
+                  .get(symbol.name)
+              ) match {
+                case Some(polySymbol) =>
```

Comment:
This may be subjective but this seems a _very weird_ way to write
```scala
val polySymbol =  mod.unsafeAsModule().getScope.getPolyglotSymbols.get(symbol.name)
if (polySymbol != null) { ... } else { ... }
```
I really like the `Option` type but here it looks like an overkill to me - especially with the multiline formatting it took me some time to correctly notice what is getting pattern matched here.